dashboard/src/reducers/kube.test.ts

```diff
@@ -44,5 +57,84 @@ describe("authReducer", () => {
         items: { foo: { isFetching: false, error } },
       });
     });
+
+    describe("openWatchResource", () => {
+      it("adds a new socket to the state for the requested resource", () => {
+        const newState = kubeReducer(undefined, {
+          type: actionTypes.openWatchResource,
+          payload: {
+            ref,
+            handler: jest.fn(),
+          },
+        });
+        const socket = newState.sockets[ref.watchResourceURL()];
+        expect(socket).toBeDefined();
+      });
+
+      it("does not open a new socket if one exists in the state", () => {
+        const existingSocket = ref.watchResource();
+        const state = {
+          ...initialState,
+          sockets: {
+            [ref.watchResourceURL()]: existingSocket,
+          },
+        };
+        const newState = kubeReducer(state, {
+          type: actionTypes.openWatchResource,
+          payload: {
+            ref,
+            handler: jest.fn(),
+          },
+        });
+        expect(newState).toBe(state);
+        expect(newState.sockets[ref.watchResourceURL()]).toBe(existingSocket);
+      });
+
+      it("adds the requested handler on the created socket", () => {
+        const mock = jest.fn();
+        const newState = kubeReducer(undefined, {
+          type: actionTypes.openWatchResource,
+          payload: {
+            ref,
+            handler: mock,
+          },
+        });
+        const socket = newState.sockets[ref.watchResourceURL()];
+        const listener = (socket as any).listeners.message[0];
```

Comment:
this is a bit weird, how is a `listener` a message of the listener? I guess that's how the library has it defined but can you add a comment explaining this?

Reply:
yes good point, as far as I can tell, `listener` comes from `mock-socket` (https://github.com/thoov/mock-socket/blob/master/index.d.ts#L7), but TS doesn't use the types from that library (I think it just uses the default WebSocket type) and I can't figure out how to use it. I'll add a comment about that.